### ISSUE TITLE
dns-01 validation fails with the gandi provider when the requested domain is a wildcard record

### ISSUE THREAD

**Issue by nmasse-itix** [xenolf/lego#472]
Hi,

When using Gandi as a DNS provider and requesting a certificate for a domain that is declared as a wildcard record in gandi, the dns-01 validation fails with:

```
GANDI_API_KEY=<HIDDEN> /usr/local/bin/lego --domains xch.app.itix.fr --email <HIDDEN> -a --dns gandi run
2018/01/18 13:22:15 [INFO][xch.app.itix.fr] acme: Obtaining bundled SAN certificate
2018/01/18 13:22:16 [INFO][xch.app.itix.fr] AuthURL: https://acme-v01.api.letsencrypt.org/acme/authz/<HIDDEN>
2018/01/18 13:22:16 [INFO][xch.app.itix.fr] acme: Trying to solve DNS-01
2018/01/18 13:22:17 [xch.app.itix.fr] Could not obtain certificates
	Error presenting token: Gandi DNS: RPC Error: (510042) Error on object : OBJECT_DOMAIN (CAUSE_NOTFOUND) [Domain 'app.itix.fr.' doesn't exist.]
```

### Diagnosis

My DNS setup is :
 - itix.fr is hosted by Gandi 
 - there is a wildcard `A` record for *.app.itix.fr:
```
*.app IN A 1.2.3.4
```

The issue comes from line 78-85 at [providers/dns/gandi/gandi.go](https://github.com/xenolf/lego/blob/cc326ce155a7916c7a784aafb3c845722a4da698/providers/dns/gandi/gandi.go#L78-L85):

```go
	authZone, err := findZoneByFqdn(fqdn, acme.RecursiveNameservers)
	if err != nil {
		return fmt.Errorf("Gandi DNS: findZoneByFqdn failure: %v", err)
	}
```

After this call, `authZone` is `app.itix.fr.`. In fact, the SOA reply from Gandi is not guaranteed to always be the zone id. When using wildcard records that have dots in their relative names, the SOA reply seems to differ from the zone id (my zone is "itix.fr"):
```console
$ dig xch.app.itix.fr soa

; <<>> DiG 9.9.4-RedHat-9.9.4-51.el7_4.1 <<>> xch.app.itix.fr soa
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 54373
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;xch.app.itix.fr.		IN	SOA

;; AUTHORITY SECTION:
app.itix.fr.		1799	IN	SOA	a.dns.gandi.net. hostmaster.gandi.net. 1516266630 10800 3600 604800 10800

;; Query time: 25 msec
;; SERVER: 62.210.189.239#53(62.210.189.239)
;; WHEN: Thu Jan 18 13:35:42 CET 2018
;; MSG SIZE  rcvd: 106

``` 

### Suggestion

Since there is no way to create Gandi zones for anything else than the base domain (xyz.tld), I suggest to modify the `gandi` provider to determine the zone by taking the last two components of the FQDN. 

What do you think ? Would it be a suitable and acceptable pull request ? 

Thanks for your help and support. :)

**Comment by nmasse-itix** [comment]
Forgot to mention the version :
```
$ /usr/local/bin/lego -v
lego version v0.4.1-0-g67c86d860a79
```

**Comment by captncraig** [comment]
The provider doesn't strictly need to use `FindZoneByFQDN`. I kinda think its a bit odd essentially looking for a zone with existing dns records to see what matches. Could just look through the zones in your gandi account and pick the one that is the longest matching suffix to the fqdn in question?

**Comment by nmasse-itix** [comment]
Yes, I think it would be more reliable.

**Comment by nh2** [comment]
@nmasse-itix Not super sure if relevant but check that in your Gandi domain settings, it doesn't say `Gandi's Classic DNS` but `Gandi's LiveDNS`.

**Comment by tchabaud** [comment]
@nh2 : Same problem here (but with Gandi v5 API), and my Gandi DNS settings are set to Live DNS.

**Comment by goldsziggy** [comment]
Same problem as well, I am using the Gandi v5 API as well.  Confirmed my DNS is set to Live DNS as well.

**Comment by ldez** [comment]
Closed by #664
